test/release/examples/primers/modules.chpl

```diff
@@ -0,0 +1,390 @@
+/*
+ * Module Primer
+ *
+ * This primer introduces the concept of modules, a strategy for encapsulating
+ * code for later use.  It covers:
+ *   how to define a module
+ *   namespace control for a module's symbols
+ *   outside access of a module's symbols
+ *   namespace control when using a module in an outside context, including:
+ *     unlimited
+ *     explicit exclusion of symbols
+ *     explicit inclusion of symbols
+ *     renaming included symbols
+ *   outside access of modules in different files
+ */
+
+
+/* A module is a grouping of code - each program consists of at least one
+   module, and every symbol is associated with some module.  If all the code of
+   a file is not enclosed in an explicit module, defined using the syntax below,
+   then the file itself is treated as a module with the same name as the file
+   (minus the .chpl suffix).
+ */
+module modToUse {
+
+  /* In this case, foo is a public global variable which lives in the module
+     modToUse */
+  var foo = 12;
+
+  // As is bar.
+  var bar: int = 2;
+
+  /* baz is a public function which lives in modToUse */
+  proc baz (x, y) {
+    return x * (x + y);
+  }
+
+  /* A module defined within another module is known as a nested module.  These
+     modules can take advantage of symbols defined within their parent module,
+     but their parent module must take additional steps to access the contents
+     of the nested module.  We will cover how to access the contents of other
+     modules later in this primer.
+  */
+  module Inner1 {
+    /* The variable foobar is utilizing modToUse's foo and bar variables. */
+    var foobar = foo + bar;
+  }
+
+  /* Note that since there is other code after the end of modToUse's definition,
+     modToUse is a submodule of the overarching file module "modules".
+  */
+
+
+
+  /* Often, the functionality of a group of code doesn't need to be completely
+     visible to outside clients.  Sometimes it would be harmful for the client
+     to access a function that isn't part of the blessed API, and sometimes it
+     just adds unnecessary clutter to the namespace.  To avoid these cases, a
+     symbol can be declared "private" - this means that only code defined within
+     the same scope as the definition of this symbol can access it.
+  */
+
+  /* Here, hiddenFoo is a private global variable, which is only accessible by
+     symbols defined within modToUse
+  */
+  private var hiddenFoo = false;
```

Comment:
Could this be moved up among the other module-scoped variables?  It seems more stylistically attractive/normal to put them together, and since they were named as being public, it makes sense to show how private things can be declared in contrast to the public things right there.  It also feels like more of a first-order concept than sub-modules.  Ditto the private function just below (which I'd sort next to the public function(s) above).